lte/gateway/c/oai/lib/itti/memory_pools.c

```diff
@@ -229,7 +235,10 @@ static inline int items_group_put_free_item(
 {
   items_group_position_t put_raw;
   items_group_position_t put;
+  items_group_index_t tmp;
 
+  /* FIXME(pboldin): we have a potential infinite loop here */
+retry:
```

Comment:
Why do we need to retry? What is the semantics of this function?

Reply:
The `get` and `put` indexes may overwrap, leaving `get` to point beside the currently empty place and this may cause 
asserts in the `put` code to fail.

I will need some more time to get through this code.

Reply:
With number of threads > num of items, I understand that duplicate allocation is possible due to get overrunning put position. But, we are operating in the opposite regime where num of items >> num of threads. In this operating regime when get overruns put position, it ends up at a free location and it rescinds. Since we cannot put more than we get, put never overruns get position.

Having said that I observed put event failure when simulating with 3K UEs and running test_attach_detach_multi_ue.py. Trying to understand why and how they happened.

Reply:
> But, we are operating in the opposite regime where num of items >> num of threads.
IIRC no you are not. Some queues are really small, like 20000 bytes has around 20 elements.

Do you have the corefiles and can you send them to me along with execs?

Reply:
@paboldin I cannot reproduce the assert failure on the latest master, but it used to generate a backtrace as follows:

Jun 10 23:22:11 magma-dev mme[3656]: Assertion (items_group->indexes[put] <= ITEMS_GROUP_INDEX_INVALID) failed!
Jun 10 23:22:11 magma-dev mme[3656]: In items_group_put_free_item() /home/vagrant/magma/lte/gateway/c/oai/lib/itti/memory_pools.c:252
Jun 10 23:22:11 magma-dev mme[3656]: Index at current put position (41472) is not marked as free (66537)!
Jun 10 23:22:11 magma-dev mme[3656]: Assertion (result == 0) failed!
Jun 10 23:22:11 magma-dev mme[3656]: In memory_pools_free() /home/vagrant/magma/lte/gateway/c/oai/lib/itti/memory_pools.c:709
Jun 10 23:22:11 magma-dev mme[3656]: Failed to free memory pool item (pool 0, item 41477)!
Jun 10 23:22:11 magma-dev mme[3656]: Assertion (rc == 0) failed!
Jun 10 23:22:11 magma-dev mme[3656]: In itti_free() /home/vagrant/magma/lte/gateway/c/oai/lib/itti/intertask_interface.c:208
Jun 10 23:22:11 magma-dev mme[3656]: Failed to free memory at 0x7f222ec5f160 (5)
Jun 10 23:22:11 magma-dev mme[3656]:  Will Intentionaly raise SEGFAULT to be catched by GDB!
Jun 10 23:22:11 magma-dev mme[3656]: Obtained 7 stack frames.

Reply:
In pool zero, the order of index numbers is in ten thousands. The smallest created pool has 100 items. The number of threads on the other hand are handful.

Reply:
The thing is that whoever receives message holds it until it is done with it, sometimes blocking in syscalls or otherwise.

This means that there may be zero free items and multiple threads are trying (and failing) to allocate more items, leading to the result that you see in the logs.

Reply:
I have further analyzed the code and it looks fine even when `get` position overruns `put` position as this condition is only temporarily.

For instance, let's say that there are no free items and two threads are trying to simultaneously allocate items.

First one is being handled with `put == get` condition and fails to allocate, decrementing `get` correctly.

Second one is being handled with `index = indexes[get]` which is equals ITEMS_GROUP_INDEX_INVALID and successfully exchange it with itself, but then the following code sees index <= ITEMS_GROUP_INDEX_INVALID and decrements `get` correctly once again.

If another thread have just freed the position we have in `get` index (e.g. two messages were freed while our thread was re-scheduled) then we either fetch correct `index` and claim the item or we are once again getting index = ITEMS_GROUP_INDEX_INVALID.

My only additional comment is that code freeing items should use atomic assignment (even without cmpxchg) so that cache coherence is correct and we are not missing any items that were just freed in another thread.

Please note that all of the above only holds as long as number of items >> number of threads.

Overall, I think the code still needs more stress testing and integration tests must be able to run in parallel (with different base IMEIs), otherwise the very first stadium event may get the code broken.